aiida/engine/processes/workchains/awaitable.py

```diff
@@ -62,6 +61,7 @@ def construct_awaitable(target):
             'action': AwaitableAction.ASSIGN,
             'target': awaitable_target,
             'outputs': False,
+            'resolved': False,
```

Comment:
Not fully sure why `resolved` used here, I guess it is used to prevent repeatedly resolve the awaitable, since there are two mechanism now. Could you add an entry in docstring for explanation?

Reply:
I based this PR on another branch that is used for PR #4154 . Since that one touched quite a bit of the same code, I didn't want to have to deal with the merge conflicts. So I just based this PR of off that branch. That is why there are 3 commits. Only the last commit is relevant to solve this issue. The other commits reinstate the second mechanism of checking awaitables, as you mentioned. That solves one problem, but added another problem with the awaitables that now, sometimes, it can be "actioned" twice, putting the same node in the context twice, which is unwanted. So I needed some kind of "guard" to make sure it is done only once. The `resolved` attribute is my solution. It has its own problems and possibilities for failure but was the best solution I could come up with for now.